src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs

```diff
@@ -1,156 +1,179 @@
+                .Replace(downstreamRoute.DownstreamPathTemplate.Value, templatePlaceholderNameAndValues);
 
-using Ocelot.Responses;
+            var downstreamRequest = httpContext.Items.DownstreamRequest();
 
-using Ocelot.DownstreamUrlCreator.UrlTemplateReplacer;
+            if (response.IsError)
+            {
+                Logger.LogDebug("IDownstreamPathPlaceholderReplacer returned an error, setting pipeline error");
+
+                httpContext.Items.UpsertErrors(response.Errors);
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(downstreamRoute.DownstreamScheme))
+            {
+                //todo make sure this works, hopefully there is a test ;E
+                httpContext.Items.DownstreamRequest().Scheme = downstreamRoute.DownstreamScheme;
+            }
+
+            var internalConfiguration = httpContext.Items.IInternalConfiguration();
+
+            if (ServiceFabricRequest(internalConfiguration, downstreamRoute))
+            {
+                var pathAndQuery = CreateServiceFabricUri(downstreamRequest, downstreamRoute, templatePlaceholderNameAndValues, response);
+
+                //todo check this works again hope there is a test..
+                downstreamRequest.AbsolutePath = pathAndQuery.Path;
+                downstreamRequest.Query = pathAndQuery.Query;
+            }
+            else
+            {
+                var dsPath = response.Data;
+
+                if (ContainsQueryString(dsPath))
+                {
+                    downstreamRequest.AbsolutePath = GetPath(dsPath);
+
+                    if (string.IsNullOrEmpty(downstreamRequest.Query))
+                    {
+                        downstreamRequest.Query = GetQueryString(dsPath);
+                    }
+                    else
+                    {
+                        //downstreamRequest.Query += GetQueryString(dsPath).Replace('?', '&');
+                        var newQueryString = GetQueryString(dsPath).Replace('?', '&');
+                        newQueryString = MergeQueryStringsWithoutDuplicateValues(downstreamRequest.Query, newQueryString);
+                        downstreamRequest.Query = "?" + newQueryString;                    }
+                }
+                else
+                {
+                    RemoveQueryStringParametersThatHaveBeenUsedInTemplate(downstreamRequest, templatePlaceholderNameAndValues);
+
+                    downstreamRequest.AbsolutePath = dsPath.Value;
+                }
+            }
+
+            Logger.LogDebug($"Downstream url is {downstreamRequest}");
+
+            await _next.Invoke(httpContext);
+        }
+
+        private static string MergeQueryStringsWithoutDuplicateValues(string queryString, string newQueryString)
+        {
+            var queries = HttpUtility.ParseQueryString(queryString);
+            var newQueries = HttpUtility.ParseQueryString(newQueryString);
+
+            var dict = new Dictionary<string, string>();
```

Comment:
```suggestion
            newQueries.AllKeys.Where(key => !string.IsNullOrEmpty(key)).ToDictionary(key => key, key => newQueries[key]);
```

Reply:
✔️ Fixed in https://github.com/ThreeMammals/Ocelot/pull/1182/commits/fef94b63b48cea1bbdd36180001e4e44d5586905